open_spiel/games/battleship.cc

```diff
@@ -0,0 +1,849 @@
+  }
+}
+
+std::string BattleshipState::ActionToString(Player player,
+                                            Action action) const {
+  SPIEL_DCHECK_TRUE(player == Player1 || player == Player2);
+
+  // FIXME(gfarina): It was not clear to me from the documentation whether
+  //     the next condition is always guaranteed.
+  //
+  //     Action ids are reused at different states, so the meaning of the same
+  //     action id is contingent on the current state.
+  SPIEL_CHECK_EQ(player, CurrentPlayer());
+
+  if (!AllShipsPlaced()) {
+    // If we are here, we still have some ships to place on the board.
+    //
+    // First, we find the first ship that hasn't been placed on the board yet.
+    const ShipPlacement ship_placement = DeserializeShipPlacementAction(action);
+    return ship_placement.ToString();
+  } else {
+    // In this case, the only thing the player can do is to shoot on a cell
+    const Shot shot = DeserializeShotAction(action);
+    return shot.ToString();
+  }
+}
+
+std::string BattleshipState::ToString() const {
+  std::string state_str;
+  for (const auto& move : moves_) {
+    if (move.player == Player1) {
+      absl::StrAppend(&state_str, "/1:");
+    } else {
+      absl::StrAppend(&state_str, "/2:");
+    }
+    if (absl::holds_alternative<ShipPlacement>(move.action)) {
+      absl::StrAppend(&state_str,
+                      absl::get<ShipPlacement>(move.action).ToString());
+    } else {
+      SPIEL_DCHECK_TRUE(absl::holds_alternative<Shot>(move.action));
+      absl::StrAppend(&state_str, absl::get<Shot>(move.action).ToString());
+    }
+  }
+  return state_str;
+}
+
+bool BattleshipState::IsTerminal() const {
+  return CurrentPlayer() == kTerminalPlayerId;
+}
+
+std::vector<double> BattleshipState::Returns() const {
+  SPIEL_CHECK_TRUE(IsTerminal());
+  const BattleshipConfiguration& conf = bs_game_->configuration;
+
+  // The description of the game in the header file contains more details
+  // about how the payoffs for the players are computed at the end of the
+  // game, as well as the meaning of the `loss_multiplier`.
+  const double loss_multiplier = conf.loss_multiplier;
+
+  double damage_pl1 = 0.0;
+  double damage_pl2 = 0.0;
+  for (const Ship& ship : conf.ships) {
+    if (DidShipSink(ship, Player1)) damage_pl1 += ship.value;
+    if (DidShipSink(ship, Player2)) damage_pl2 += ship.value;
+  }
+
+  return {damage_pl2 - loss_multiplier * damage_pl1,
+          damage_pl1 - loss_multiplier * damage_pl2};
+}
+
+std::unique_ptr<State> BattleshipState::Clone() const {
+  return std::make_unique<BattleshipState>(*this);
+}
+
+std::string BattleshipState::InformationStateString(Player player) const {
+  // XXX(gfarina): It looks like `api_test.py` wants to see the substring
+  //     "player >= 0" when the method fails because an invalid player was
+  //     supplied. That's why the redundant check is appended to the end.
+  SPIEL_CHECK_TRUE(player >= Player1 && player <= Player2 && player >= 0);
```

Comment:
SPIEL_CHECK_GE(player, 0);
SPIEL_CHECK_LT(player, NumPlayers());

Reply:
So, in an ideal world, this would be

```
SPIEL_CHECK_TRUE((player == Player1) || (player == Player2));
```

Unfortunately OpenSpiel is not being super collaborative about that. I guess that could become the intersection of the two constraints you were proposing, though I dislike to have to go down to the machine representation of the player type.

Reply:
I think in an ideal world it would be
```
struct Player {
  int p;
}
```
to avoid type aliasing with int. If you're familiar with Haskell, the difference between `type Player = Int` and `newtype Player = Player Int` (https://wiki.haskell.org/Type#Type_and_newtype). 

But doing this is too much of a change in the codebase I'm afraid.

Reply:
Ah yeah I am a fan of "hard" newtypes too. Given that we are already moving from `Player1` to `Player{0}`, I think at this point we can just go with

```
SPIEL_CHECK_TRUE((player >= 0) && (player < NumPlayers()));
```